apps/src/code-studio/components/TeacherContentToggle.js

```diff
@@ -0,0 +1,116 @@
+/* globals appOptions */
+
+import React, { PropTypes } from 'react';
+import Radium from 'radium';
+import { connect } from 'react-redux';
+import { ViewType, fullyLockedStageMapping } from '../stageLockRedux';
+import { isHiddenFromState } from '../hiddenStageRedux';
+
+const styles = {
+  hidden: {
+    display: 'none'
+  }
+};
+
+// TODO - unit tests
+// TODO - eyes tests
+/**
+ * When viewing a puzzle, we want teachers to be able to toggle between what the
+ * student would see and what they see as a teacher. In some cases (such as
+ * locked stages and hidden stages) this means hiding the main content, and
+ * replacing it with something else.
+ * We accomplish this by having the server render that other content to a known
+ * dom element (#locked-stage, #hidden-stage). This component then creates
+ * container elements for the main content and any other content, and toggles
+ * which of those containers is visible as appropriate.
+ */
+const TeacherContentToggle = Radium(React.createClass({
+  propTypes: {
+    viewAs: PropTypes.string.isRequired,
+    hiddenStagesInitialized: PropTypes.bool.isRequired,
+    sectionsAreLoaded: PropTypes.bool.isRequired,
+    isHiddenStage: PropTypes.bool.isRequired,
+    isLockedStage: PropTypes.bool.isRequired
+  },
+
+  componentDidMount() {
+    // Show this element, as parent div (refs.lockMessage) now owns visibility
+    $('#locked-stage').appendTo(this.refs.lockMessage).show();
+    $('#hidden-stage').appendTo(this.refs.hiddenMessage).show();
+    // Server initially sets level-body opacity to 0 when viewAs=Student so that
+    // student view doesnt show content while we make async calls. Once this
+    // component has mounted, we move level-body into our first div, which will
+    // now own toggling visibility
+    $('#level-body').css('opacity', '').appendTo(this.refs.content);
+    console.log('componentdidmount');
+  },
+
+  render() {
+    const { viewAs, hiddenStagesInitialized, sectionsAreLoaded, isLockedStage, isHiddenStage } = this.props;
+
+    const frameStyle = {
+      position: 'relative',
+      zIndex: 1,
+      backgroundColor: getComputedStyle(document.body).backgroundColor,
+      height: window.screen.height
+    };
+
+    let contentStyle = {};
+    let hasOverlayFrame = (isLockedStage || isHiddenStage);
+
+    if (viewAs === ViewType.Student) {
+      // Keep this hidden until we've made our async calls for hidden_stages and
+      // locked stages, so that we don't flash content before hiding it
+      if (!hiddenStagesInitialized || !sectionsAreLoaded || hasOverlayFrame) {
+        contentStyle.opacity = 0;
```

Comment:
I didn't think to try visibility: hidden. I can do that. I know with display: none, droplet and blockly would end up looking weird if they started out hidden and later got displayed.